Ignition  and Fuel schedulers

TBAMax opened the pull request:
Upgraded ignition and fuel schedulers.
This PR includes:
1. Get fuel and ignition schedules to work as originally intended: dwell timing and fuel injection pulsewidth directly applied to the timers. Apparently there were some computational hardships and problems with having all integer types correct, to have integer math on the timers work correctly (regarding need to take account that all math must work correctly with timer counter that rolls back to 0 periodcally). That is all honed out by now.
2. No change in loop speed on atmega2560 or only a very slight drop.
3. Noticeable dwell timing accuracy improvement, observeable with logic analyser. No observeable random extra ignition events visible during testing. Tested with all speed ranges from 60rpm to 16000rpm.
4. Smooth, no misses, no glitches, transition from cranking to running ignition dwell and timing tested with simulator and logic analyser.
5. Replaces #727 and #578 and #653 , related to #686 
6. New ignition mode also working and tested. With 60-2 trigger (harder case) seems to be working fine.
7. on Atmega3560 compared to master:133 Bytes less RAM used, 4324 bytes less flash used.
8. Updated ignition dwell limiter.
  update information 3.2023: Dwell limiter works also against possible race contitions when port register read-modify-write 
  type writes are done in some other parts of the code. Fixed possible race condition in dwell limiter itself when next ignition pulse falls within dwell limiter range, to prevent dwell limiter affecting the next ignition pulse at high RPM.
9. New Unity tests for the new schedulers (468 tests total for schedulers on atmega2560, more for boards that have more channels)
10. (update information 3.2023) Fuel injection duty cycle range extended to 100% (previously limited to 95%). This means continuous ON fuel flow from injectors at extreme cases.

TBAMax committed:
New optimized fuel and ignition schedules.

Some explaining remarks:
1.Separate dwell limiters is no longer needed because of rooted out problems with schedulers. Now schedulers itself act as dwell limiters.
2. Replace fixed cranking ignition functions also with new schedulers (in the decoders.ino).

TBAMax committed:
update code comments in scheduler.h

pazi88 commented:
Sounds promising 🙂

noisymime commented:
I've added comments for a few things I've spotted whilst reviewing this, mostly where something is fundamentally changed from the current functionality. 

Have tested through some of the most common setups and things generally seem to be OK. The resulting code is certainly cleaner, but there's quite a few areas where edge case checks have been removed, which makes me a little nervous. I'll need to do some more testing around these though

Here's a few more general comments anyway:

* The cleanup of the scheduler code is certainly overdue, but does come with some (Probably inevitable) downsides such as the increased time spent within ISRs and the additional context switching that will be required due to the extra function calls. The scheduler code in particular sees quite a number of function calls added to the ISRs, which is something that will need to be considered (Noting that the calls won't be inlined)
* Pretty much all the unit testing is broken in this PR as it stands at the moment. This will need to be fixed before pulling this in and obviously all the checks will need to be verified also. 
* I'm seeing some weirdness on the fuel pulses at high rpm/load (Beyond the duty cycle limiter) where pulses are either not starting or not completing correctly. Need to dig into this more though.

TBAMax commented:
@noisymime thanks for the comments. Today I managed to reorganize things so that I got rid of the small helper functions, and also got rid of the need for using the function pointers to the timer helper functions in the structs.
  The result is that there is less calls in the interrupt code (checked that from the disassembly listing). Also the code is even more clean now. And still can use the generic function for dealing with the schedulers. 
  The good thing is that now calls(previously) are inlined as much as possible, so I consider there is no increased time in the ISR from this PR at the moment. This all for today. Looked with logic analyser, all seems to be operating as expected at least on atmega. 
  Fuel duty cycle limiter behaviour still needs to be adressed and unit tests.

noisymime commented:
@TBAMax Re-reading my comment from the other day and I realised it probably came across a bit negative. Wasn't my intention as this is a really good PR. Just want to make sure it's not causing any issue when pulled in given it's size and how it's touching on so many critical areas.

TBAMax commented:
@noisymime no problem. I hope I did not ruin the PR with the latest addition of structs with virtual functions. I researched into it and it still seems good solution in my eyes. Specially for the solution against having function calls that can not be inlined. With having virtual functions in the main struct and then overriding them in the derived child struct(class) allows those functions to be inlined, and it is being confirmed by looking at reduced amount of instructions (and icalls) in the dissassembly listing. So functionally and reduced codewise it looks good, but it is essentially not so conservative C coding, but rather C++.

TBAMax committed:
Extend fuel Duty cycle limiter to 100%

Allow for 100%duty cycle on fuel injectors. This also allows fuel schedulers to have overlapping pulses without major problems. During overlapping pulses injectors are simply held open all the time.

noisymime committed:
Merge branch 'master' of https://github.com/noisymime/speeduino

noisymime committed:
Fix issue with SD RPM threshold not correctly detecting on/off

adbancroft commented:
@TBAMax I may have a way to remove the virtual function calls. Is this PR still being actively worked on?

@noisymime Is this PR likely to be accepted?

TBAMax commented:
> @TBAMax I may have a way to remove the virtual function calls. Is this PR still being actively worked on?

@adbancroft Bring on the info about alternative ways! Currently this PR is included to the https://github.com/TBAMax/speeduino/tree/mainbranch (with some other stuff) for testing (some lag of maintaining mergeability here). And I have been using it on the car to get the feel how it is behaving. Although some recent changes about the full/semi sequential switching not included there also. It was little hard to understand why some of the scheduler swithing was moved to the init.ino recently, and I need to chew that logic out yet, to bring this up to date.

adbancroft commented:
@TBAMax How would I bench test changes using ArduStim? Is there anything in TunerStudio to log?

adbancroft commented:
@TBAMax I'm working on a patch for you. Question: the new fuel scheduler adds a new check when setting the next*Compare members: there has to be 400 timer ticks remaining in the current injection pulse. Why is this needed?

TBAMax commented:
> @TBAMax I'm working on a patch for you. Question: the new fuel scheduler adds a new check when setting the next*Compare members: there has to be 400 timer ticks remaining in the current injection pulse. Why is this needed?
@adbancroft This was needed to make 100%duty cycle possible, to cater every possible case.
 Indeed the hardcoded constant is not nice. Maybe it can be calculated as following:
`if ((timeout < MAX_TIMER_PERIOD) && ((COMPARE_TYPE)(targetSchedule->endCompare-targetSchedule->getFuelCounter())>(uS_TO_TIMER_COMPARE(INJECTION_REFRESH_TRESHOLD)))`
need to give it a thought, and then test everything again...
When injection pulses overlap (injector always "ON" condition), then fuelchedule status is always RUNNING so all schedule setting in the main loop is actually done via nextSchedule. So the nextschedule update can not also be too close to the interrupt. Maybe it can be done simpler also. I was observing the duty cycle limit. It was previously set capped to 95% because schedulers could not take 100% and would have quite random behaviour at more than 95% (because of fluctuation of 'timePerDegree' causing overlap sometimes). So this was what I came up for now.

noisymime commented:
Most of the questions I posted above look like they still stand, even after the changes. Some look like they might just be copy/paste type things, but others might be a bit more fundamental.

TBAMax commented:
Just a note here that recently  merged  changes from #876 probably affect this too, because ignition or fuel timer macros are changed there. So when rewriting this at some future point, need to look at those changes there.
Currently I still consider the ignition and fuel schedulers presented here to be superior to those of the current speeduino/master. But it is not possible to maintain such a big change reliably as a mergeable PR, as it is practically maintaining a separate fork. So fresh rewrite starting from the master is needed for this, to avoid accidentally reversing some bugfixes that have been done meanwhile.

adbancroft commented:
@TBAMax Teemo - I agree with this approach to the timers and the need to recode against latest master. What can I do to help?

A different branching strategy could make maintaining this easier: in your separate branch in your repo, *rebase* from master regularly. This keeps the both commit histories easily visible, rather than bulk merges that contain multiple commits under one "Merge branch 'master' into IgnAndFuel" comment. It may complicate in other ways though.

TBAMax commented:
> @TBAMax Teemo - I agree with this approach to the timers and the need to recode against latest master. What can I do to help?
> 
> A different branching strategy could make maintaining this easier: in your separate branch in your repo, _rebase_ from master regularly. This keeps the both commit histories easily visible, rather than bulk merges that contain multiple commits under one "Merge branch 'master' into IgnAndFuel" comment. It may complicate in other ways though.

@adbancroft  Some points with that was that it breaks the automated (unit)tests for the schedulers. I was uncertain how to deal with that. Either make new tests, or try to modify the code so that it can be squeezed to satisfy the existing testing pattern..., or combined both.
Other issues I consider pretty much fixed (irregular pulses beyond the duty cycle limiter, function calls in the ISR inlining).

TBAMax commented:
Overdwell protection currently under inspection here.
Objectives: 
1.Gather all information about current status of the overdwell protection and its working mechanisms.
2.If missing pieces or places of possible improvement found, make nessesary changes.
3.Update documentation of the PR about the topic.

noisymime commented:
Please do look at the other comments I made in the review above. These may have been addressed already, but there were a few things there that looked questionable.

TBAMax committed:
Updates from the recent rebase

Some housekeeping from the most recent rebase work.

TBAMax committed:
Fix accidentally corrupted SD_logger.ino

For some reason SD_logger.ino file had not been correctly up to date with the current master. Now just copied the fresh file from the master.

TBAMax commented on speeduino/speeduino.ino:
```diff
@@ -890,399 +844,134 @@ void loop(void)
 #if IGN_CHANNELS >= 2
-        if (maxIgnOutputs >= 2)
+        if (maxIgnOutputs >= 2 && !BIT_CHECK(curRollingCut, IGN2_CMD_BIT))
         {
```
Something is strange with the rolling cut here! Why the master do not have it any more?

TBAMax replied:
Ok, checked that: change view picks that up from different place. In reality there is no issue. Rolling cut is still in place. So all ok with that.

TBAMax commented on speeduino/speeduino.ino:
```diff
@@ -1424,7 +1111,7 @@ byte getAdvance1(void)
     //Alpha-N
-    currentStatus.ignLoad = currentStatus.TPS * 2;
+    currentStatus.ignLoad = currentStatus.TPS;
 
```
Seems like accidental reversal of totally unrelated thing! How can this happen?

TBAMax replied:
Fixed that in the nex commit.

TBAMax committed:
Resolve some more left conflicts with master.

Some updates on the master were gone unnoticed and in the treat of being reversed in this branch.
Also crankRPM setting was updated twice , once in the main loop and once in the interrupt. Little unrelated topic but removed the interrupt on in here.

DeionSi commented on speeduino/scheduler.h:
```diff
@@ -44,247 +33,247 @@ See page 136 of the processors datasheet: http://www.atmel.com/Images/doc2549.pd
+
+//The ARM cores use seprate functions for their ISRs
 #if defined(ARDUINO_ARCH_STM32) || defined(CORE_TEENSY)
-  static inline void fuelSchedule1Interrupt(void);
```
The argument should probably be void on these. Some MISRA rule I believe.

TBAMax replied:
Yeah, fixed in the next commit here.

TBAMax committed:
Add "void" to function declarations and definitions

TBAMax committed:
Reintroduce dwell limiter

Brand new highly optimised ignition dwell limiter actually.

DeionSi commented:
The tests are no longer passing. It seems they need to be updated.

TBAMax commented:
> The tests are no longer passing. It seems they need to be updated.

Is there a technical reference of how to run those tests?

DeionSi commented:
> Is there a technical reference of how to run those tests?

https://docs.platformio.org/en/stable/advanced/unit-testing/index.html

In vscode platformio
![image](https://user-images.githubusercontent.com/26184874/214609028-c8dc9e80-8472-4a9e-a2b9-64607b57a529.png)

DeionSi commented:
Have a look at @adbancroft 's [PR](https://github.com/TBAMax/speeduino/pull/11). It seems to have a lot of testing stuff in it.

TBAMax committed:
Merge branch 'noisymime:master' into IgnAndFuel

TBAMax committed:
Rename variables for better readability

Rename timeout to endTimeout, in effort to make it more clear that it is used as time left for pulse end.
No functional changes, just renaming and correcting some comments.

TBAMax committed:
New tests for schedulers.

TBAMax committed:
Upgrade description of schedules in the files.

No functional changes. Only cleanup.

TBAMax commented:
Update

@DeionSi @adbancroft  I have now upgraded the tests here , and managed to run them on atmega physically also. I must say this is quite fun! Took the more advanced route and included the crankmath into the test, so the schedules can be tested at various RPMs.

@noisymime, when looking over the checklist now : 
- The function calls in the interrupts are minimized. The virtual functions are here to stay. I looked the alternative of using function pointers or references to the `setCompare(x) `and  `getTtimer()`, this was not good, and resulted in indirect function calls from the interrupt routine vs. with virtual function style as now all is inlined.( `avr-objdump -d -S firmware.elf >firmware.lst`)
  So all that is left from the function calls inside the interrupt routine is `startFunction()` and `endfunction()` (indirect calls) but nothing to do about those because we need to be able to reconfigure what those point to. 
- Dwell limiter is now  fixed and reintroduced. Though there is possibility to add corresponding unity test for it in the future. Current present day status of this is that I have not tested it on the real hardware.
- Fuel schedules performance should be resolved with the allowing of the PW up to 100%. So that it do not have any more weirdness at near or even over 100% duty cycle. I remember doing extensive testing on those with the real board and Ardustim, so it should be fine now.

So thats all for today. Thanks for watching! Remember to click "Like" and "Subscribe"!

TBAMax committed:
Merge branch 'master' into IgnAndFuel

adbancroft commented:
@TBAMax 
> The virtual functions are here to stay. I looked the alternative of using function pointers

I ran a quick test and the references/function pointers solution saved 54 bytes and was 2 to 3% faster loop/sec (Ardustim, 60-2 missing tooth, multiple combinations of ignition configurations). I submitted a PR you can test.

Also, I'm seeing some ignition drop outs:
![image](https://user-images.githubusercontent.com/13982343/216494810-453cd6c5-2c34-4426-b623-4f142be53f26.png)

TBAMax commented:
> I ran a quick test and the references/function pointers solution saved 54 bytes and was 2 to 3% faster loop/sec (Ardustim, 60-2 missing tooth, multiple combinations of ignition configurations). I submitted a PR you can test.

Ok, the 54 bytes itself is not critical. But 2 to 3% loops/sec? it should be not noticeable. Possible that it comes from the counter and compare used as pointers to register. I have to check that out, seems good optimization! Timer start and timer stop are still funcitons and so make function calls, but those are not actually used regularily when engine is running.
> 
> Also, I'm seeing some ignition drop outs:

The drop outs are more serious issue to address. What is the configuration of ignition shown in the picture? It seems coils 1,3,4 firing simultaneously and only 2 is shifted? Drop outs may be from the new dwell limiter...

TBAMax commented:
> 60-2 wheel, rising edges, wasted spark, old ignition mode, Ardustim at 4K RPM.

When you turn off the dwell limiter do you see the dropouts dissapear?

TBAMax commented:
> Also, I'm seeing some ignition drop outs:

It is a dwell limiter running into the next impulse. This is possible due to he following:
This is due to the nested interrupts, the scheduler interrupt happening inside the timer2 interrupt because the timer2 interrupt is with the argument ISR_NOBLOCK
So the scheduler interrupt happens right in the middle of the dwell limiter. I originally missed to notice that there is nested interrupts on atmega...
When the scheduler interrupt happens right in the middle of dwell limiter, between the limiter comparison and limiter activation(call to the EndFunction), the new pulse gets disabled immediately by the limiter. So kind of a race condition. @adbancroft good that you observed carefully and noticed!
So now this is a known issue. Only to figure out now how to resolve this.

TBAMax committed:
Fix race condition in dwell limiter

Due to the nested interrupts there was a race condition between the dwell limiter and schedule interrupts. This caused dwell limiter to occassionally "limit out" new incoming pulses.

TBAMax commented:
I have pushed the fix for the dwell limiter. Feel free to check out!

TBAMax committed:
Revert "Fix race condition in dwell limiter"

This reverts commit de1c280cfe753941bcafc261df1646db22207239.

TBAMax committed:
New fix for the dwell limiter race condition

TBAMax commented:
Oh crap it did not work on stm32. Because cli() and sei() is not applicable there...
So new version.

VitorBoss commented:
> Oh crap it did not work on stm32. Because cli() and sei() is not applicable there... So new version.

The helper functions have their relative substitutes, if it didn't work isn't because of that.
Keep in mind ARM cores have 15 levels of interrupt nesting.

DeionSi committed:
Use existing channelDegrees for injector schedules

DeionSi commented:
Some memory and flash savings in https://github.com/TBAMax/speeduino/pull/13

TBAMax committed:
Merge pull request #13 from DeionSi/TBAMax_IgnAndFuel-channelDegrees

Use existing channelDegrees for injector schedules

TBAMax committed:
Merge branch 'master' into IgnAndFuel

TBAMax committed:
Merge branch 'master' into IgnAndFuel

TBAMax committed:
Merge branch 'noisymime:master' into IgnAndFuel

TBAMax committed:
Merge branch 'noisymime:master' into IgnAndFuel

TBAMax committed:
Merge branch 'noisymime:master' into IgnAndFuel

noisymime committed:
Allow fuel staging up to 8 total fuel channels (All reasonable combinations)

TBAMax committed:
Merge branch 'master' into IgnAndFuel

TBAMax committed:
Final clean of recent merge conflicts.

Revert most of the #1010 because it seems incopatible with most of the new schedulers here.

TBAMax committed:
Fix some invalid config.

Fix invalid config. That came in with the addition of 8 channel staging. It was that there was attempt to use non existant fuel channel and consequently the build failed.

TBAMax commented:
> I ran a quick test and the references/function pointers solution saved 54 bytes and was 2 to 3% faster loop/sec (Ardustim, 60-2 missing tooth, multiple combinations of ignition configurations). I submitted a PR you can test.

Just a note here that I have analysed the references/function pointers solution. The 54 bytes are flash not RAM. In the interrupts the references/function pointers version uses about two atmega instruction cycles more than virtual functions version, so practically same. But I did not rush to change yet, because before version is well tested and I do not have the recources at the moment to retest extensively. So I feel keeping this in mind, and when needed, do the change in the future. As for the 2% to 3% loops/sec I do not care at the moment, seems most likely in the range of possible placebo effect, as I can not really see looking at the generated asm where it would come from. Also as for the loops/s I already have some more possibly powerful ideas how to boost that (most significantly by reducing the unnessesary schedule recalculations/refreshes). But I will not add that also at the moment because it is again more of a topic for separate pull reguest to be done after (wanting to push that to the next more extensive testing session).

DeionSi commented:
@TBAMax Please see my reviews above from last month.

TBAMax commented:
> @TBAMax Please see my reviews above from last month.

@DeionSi do you mean the "Some memory and flash savings in https://github.com/TBAMax/speeduino/pull/13"  (channelDegrees for injector schedules)? That one I did pull in, and is included at the moment. It was all logical step.

DeionSi commented:
> > @TBAMax Please see my reviews above from last month.
> 
> @DeionSi do you mean the "Some memory and flash savings in [TBAMax#13](https://github.com/TBAMax/speeduino/pull/13)" (channelDegrees for injector schedules)? That one I did pull in, and is included at the moment. It was all logical step.

@TBAMax I can't find a way to link to those reviews but they are right after https://github.com/noisymime/speeduino/pull/804#issuecomment-1416138142

TBAMax committed:
Merge branch 'noisymime:master' into IgnAndFuel

TBAMax committed:
Merge branch 'noisymime:master' into IgnAndFuel

TBAMax committed:
Merge branch 'noisymime:master' into IgnAndFuel

TBAMax committed:
Merge branch 'noisymime:master' into IgnAndFuel

TBAMax committed:
Upgrade scheduler Unity tests

Upgrade scheduler unity tests to also accomodate smaller timers of Teensy 4.1

TBAMax committed:
Merge branch 'noisymime:master' into IgnAndFuel

TBAMax committed:
Fix merge conflict.

TBAMax committed:
Merge branch 'master' into IgnAndFuel

TBAMax committed:
Merge branch 'noisymime:master' into IgnAndFuel

TBAMax committed:
Merge branch 'master' into IgnAndFuel

TBAMax committed:
Merge branch 'master' into IgnAndFuel

TBAMax committed:
Merge branch 'noisymime:master' into IgnAndFuel

TBAMax committed:
Merge branch 'master' into IgnAndFuel

TBAMax committed:
Revert "Merge branch 'master' into IgnAndFuel"

This reverts commit 65ed78d9a9bd8bf4de8846364b04e795c05bd4a0, reversing
changes made to d3e5b8f3167732232bab9dfac413e7cc4da93eb0.

TBAMax commented:
It's been quite a lot of time again. I have done some merge commits here to keep this level with the speeduino:master.
So that there should be no major issues with testing it without changing the speeduino.ini. Should be able to test this without any configuration change -- just compile this and upload to the board, and it should work without any more changes compared to the speeduino:master. 
  Only thing is when you do not change the speeduino.ini then you can not set the fuel duty cycle to the 100% and test that. Without switching the `speedino.ini` it allows only up to 95% (same as master).
  
  Recent 
 
> Revert "Merge branch 'master' into IgnAndFuel"
Was simply because I decided to completely not include the recent test changes for the old schedulers on the master. Just for clarificatioin here.

TBAMax commented:
Trying to find time to build up simulator again for the Atmega and STM32 to test this after this long period of just merging the changes from master. To see if this all still works as expected. No intention to build Teensy test rig because I still do not have any Teensy board. There have been notice by @noisymime that this do not work on Teensy at the moment for some odd reason and the Teensy board just hangs on startup, but I have got the impression that it is a somewhat not related to the topic of this PR.

TBAMax commented:
Possibly the reported problems on Teensy 4.1 when the board hangs on startup, are related to nested(or not nested) interrupts.
Currently seems that using nointerrupts() and interrupts() in overdwell protection in timers.ino might be the culprit. Because it is unclear how the teensy nested interrupt controller reacts to that.

TBAMax committed:
Use ATOMIC_BLOCK instead of 'interrupts()'

Replace noInterrupts();interrupts(); pair in the timers.ino with ATOMIC_BLOCK solution. This way we do not accidentally enable nested interrupts on some platforms where interrupt nesting might not be used.

TBAMax committed:
conditional ATOMIC_BLOCK in the overdwell protection.

stm32 do not have a atomic block library, but also do not need that at the moment here. So disable atomic block for stm32 at the moment.

TBAMax committed:
Merge branch 'master' into IgnAndFuel

TBAMax committed:
Merge branch 'master' into IgnAndFuel

TBAMax committed:
Merge branch 'master' into IgnAndFuel

noisymime commented:
Current plan is for this to be one of (if not the) first changes merged in the next dev cycle. That will give a long cycle for testing of this given how large the change is to such a significant piece of the code.

TBAMax committed:
Some tweaks and fixes from testing.

-Increase INJECTION_REFRESH_TRESHOLD to avoid missing pulses at rapid RPM changes. 1.6ms is found to be optimal value.
-Fix cranking ignition.
-Rearrange interrupt blocking slightly to avoid possible scheduler state change during update.

TBAMax committed:
Merge branch 'master' into IgnAndFuel

adbancroft committed:
Test start/end callback assignment (#1044)

TBAMax committed:
Merge branch 'master' into IgnAndFuel

TBAMax committed:
Improve Dwell limiter operation

Improve dwell limiter during halfSync condition:
Dwell limiter was cutting some pulses short during halfSync wasted spark operation.
Improve dwell limiter during per tooth ignition. Add some extra checks: Max timer period, and ignition refresh treshold.
Add protection against cutting dwell too short.
Add overdwell and dwell correction protection during per tooth ignition.

TBAMax committed:
Merge branch 'master' into IgnAndFuel

TBAMax committed:
Merge branch 'master' into IgnAndFuel

TBAMax committed:
Fix some compiler warnings about variable types

Fixes compiler warnings about unsigned to signed comparison.
Some of the issues were intoduced here from the dwell limiter commits.
And some came in with the latest merge from the speeduino/master, with the new version of RPM limiter.
Good to have those fixed because those might have caused real issues later.

adbancroft commented:
I'm seeing a 5 to 7% slowdown versus the current master? 

Also, some unit tests don't compile & some fail.

TBAMax committed:
Merge branch 'noisymime:master' into IgnAndFuel

TBAMax committed:
Upgrade fuel schedule tests.

-Include INJECTION_REFRESH_TRESHOLD into the tests. That was previously not taken into account and now caused some of the tests to fail.
-force "Ignore" some cases when calculation goes into the next engine cycle, because there is no correct timing data table for those cases at the moment.

TBAMax commented:
> I'm seeing a 5 to 7% slowdown versus the current master?
> 
> Also, some unit tests don't compile & some fail.

@adbancroft thanks for the review! I fired up the tests, and indeed there were some fails. Some injection schedule tests failed. Turns out the fails were caused by partly by recently changing the `INJECTION_REFRESH_TRESHOLD`, and also for some tests `CRANK_ANGLE_MAX_INJ` was undefined. `INJECTION_REFRESH_TRESHOLD` was not accounted for in the tests and this caused the fails.

`INJECTION_REFRESH_TRESHOLD` I found by testing it is better to have larger than it was, because of possiblity of the scheduler jumping over the cycle when there are fast changes in RPM. And so may cause some fuel cycles to be skipped. (This probably applies to the current master also, not only in this PR). Anyway I included this change in at the moment.

Can not say anything about the claimed slowdown at the moment. Assuming it is on Atmega2560 that you are observing. Only thing that may cause some extra computational load, that I can think of at the moment, is addition of up to 100% fuel duty cycle capability and associated calculations.

TBAMax committed:
Fix some compiler warnings

'warning: comparison between signed and unsigned integer' FIXED

adbancroft commented:
> Assuming it is on Atmega2560 that you are observing.

Yes, I'm running on an Mega 2560: missing tooth, 60-2, crank primary trigger @ 4000 RPM on Ardustim. Consistent slow down across wasted & sequential ignition, with per-tooth timing on or off.

TBAMax committed:
Fix tests.

Some tests were left out from this branch, so bring them back in. Also test folder structure was outdated, level this with the master.

TBAMax committed:
Add shedules channelDegree initialisation

Previously it was assumed that each ignition and fuel channels initialisation is done dependin on the setup. But turns out some setups depend on initial values set to 0 and do not double set that. So this is why it is needes to initialise all channels to 0 degrees.

TBAMax committed:
Tweak tests.

Add check to prevent testing channelDegrees on channels that are inactive and so not are reguired to be initialised.
Also increase startup delay for decoders test because it seems to be more stable this way, at least with my particular board. With less delay some first tests tend to go missing or be replaced with something else.

TBAMax commented:
> > Assuming it is on Atmega2560 that you are observing.
> 
> Yes, I'm running on an Mega 2560: missing tooth, 60-2, crank primary trigger @ 4000 RPM on Ardustim. Consistent slow down across wasted & sequential ignition, with per-tooth timing on or off.

Would You mind including what are using to measure the slowdown? I mean probably TS loops/s? And what exact values are You getting?

TBAMax commented:
I am tempted to do some optimizations: Particularily modify  `triggerSetEndTeeth_missingTooth()` or `triggerSetEndTeeth` in general, to only calculate new tooth angles in case the ignition advance has changed. This should speed up the loop speed nicely. Of course this actually is not correctly in the scope of this PR.

adbancroft commented:
> Would You mind including what are using to measure the slowdown? I mean probably TS loops/s? And what exact values are You getting?

Yes, I'm comparing TS loop/s. Here's what I have:
```
Spark Setup	Per Tooth Timing	Master 2023-06-11	TBAMax	Delta
Sequential	Yes			900			835	-7%
Sequential	No			1040			970	-7%
Wasted		Yes			980			925	-6%
Wasted		No			1155			1095	-5%
```

adbancroft commented:
> Particularily modify triggerSetEndTeeth_missingTooth() or triggerSetEndTeeth in general, to only calculate new tooth angles in case the ignition advance has changed.

FYI, there's an piece of commented out code doing exactly that: speeduino.ino, line 956 on master.

TBAMax commented:
> Yes, I'm comparing TS loop/s. Here's what I have:

@adbancroft Thanks for the data. I have now verified. Indeed there is the slowdown of the described amount. I have also determined the cause: The cause is how often the `angleToTime()` function is used in the setIgnitionShedule. It is needlessly calculated for some cases. I also already have determined the possible fix, but I want to run a logic analyser test first, it needs some time to be set up again.

TBAMax committed:
Upgrade loop speed.

Some upgrades to get faster loop times.
Reposition endTimeout calculation, so that it do not get recalculated twice depending on the shedule status.
Also fixate some variable types more firmly to uint32_t.
Fix one comparison with MAX_TIMER_PERIOD, this do not improve speed but is for extra safety.

TBAMax commented:
I tested now with the logic analyser that all works as expected (on atmega2560), and pushed some upgrades to get better loop speed. Now it should fall within about 2% compared to the current moment master regarding to loops/s.

noisymime committed:
Fix issue where staging fuel channels would be active even when not required

noisymime committed:
Add unit tests for staging and update other tests

noisymime committed:
Standardise all the fuel schedule setting channel calls

TBAMax committed:
Merge branch 'master' into IgnAndFuel

TBAMax committed:
Fix additional merge conflicts

Fix additional merge conflicts that came in from the latest merge of the speeduino/master

TBAMax commented:
Apparently the updates on the master gone different route. Since this branch is well tested and working well, and functionally superior to the current master, I am keeping this updated, so that it is functionally compatible with the master. At least until changes regarding schedulers on the speeduino/master are reaching at least seemingly finished state, and can be verified against this PR (branch) and then decide if I keep maintaining this further or not.

TBAMax committed:
Merge branch 'noisymime:master' into IgnAndFuel

TBAMax committed:
Changes to the Dwell protection

Make it check same things as master
Slightly speeds up the program also. Check for status RUNNING prevents the EndFunction to be called needlessly between the pulses.

TBAMax committed:
Merge branch 'noisymime:master' into IgnAndFuel

TBAMax committed:
Merge branch 'master' into IgnAndFuel

TBAMax committed:
Merge branch 'master' into IgnAndFuel

TBAMax committed:
Merge branch 'noisymime:master' into IgnAndFuel

TBAMax committed:
Merge branch 'master' into IgnAndFuel

TBAMax committed:
Fix additional merge conflict.

There was missing conditional compilation for fuel and ignition schedules in recently added function, that is needed in this branch (Atmegas have fewer of them, and this saves some RAM)

TBAMax committed:
Initialise schedulers with constructor

Remove initialiseSchedulers() function.
(Also saves 10 bytes of RAM and 52 bytes of flash)

TBAMax committed:
Optimise fuel schedule interrupt

No need to set the schedule status again to the same value that it already is.
(also saves 74 bytes of flash and speeds up the interrupt servicing)

TBAMax committed:
Merge branch 'noisymime:master' into IgnAndFuel

TBAMax committed:
Merge branch 'noisymime:master' into IgnAndFuel

TBAMax committed:
Merge branch 'noisymime:master' into IgnAndFuel

TBAMax commented:
I did a short code rewiew at the moment to determine if some of the issues that this PR adressed still stand on the master at the moment. 
The findings:
  Seems like the main motivation that was behind this PR still stands on speeduino/master: To improve ignition timing accuracy  by not calculating the ignition moment via the "startAngle". 

  Situation is that ignition moment is calculated using the following pathway:

- [ ]  Ignition angle(from the table) -> 
- [ ] IgnitionStartAngle(angle domain math). Using engine speed, and round down to the degree accuracy, those are main sources of introduced extra inaccuracies: fluctuating speed reading and subsequent rounding down. ->
- [ ] calculate ignition moment(time domain) from IgnitionStartAngle(angle domain) (using engine speed, and current position, this is second time engine speed is used in the calculations, so any error it has is doubled!)

 Better way to calculate ignition moment (introduced in this PR) is:

- [ ] Ignition angle(from the table) ->
- [ ] Ignition time aka ignition moment(time domain). Calculated using engine speed and engine current position: least amount of computational inaccuracies!
- [ ] back calculate ignition start time(time domain). Using ignition time and dwell.

Same basically applies to the fuel injection timing.
This was the main motivation behind this PR and seems the work analysing and testing done still stands and may have some value to it.
